integrate Agent Connect

lebaudantoine opened the pull request:
### Description

https://github.com/betagouv/pad.numerique.gouv.fr/assets/45729124/ec662252-3e32-437d-8063-2b5df04883fd


### Steps

Please add the following configuration to your docker compose stack:
```
  app:
    ...
    environment:
      - CMD_AGENT_CONNECT_PROVIDER_NAME=Agent Connect
      - CMD_AGENT_CONNECT_CLIENTID=** ask me **
      - CMD_AGENT_CONNECT_CLIENTSECRET= ** ask me **
      - CMD_AGENT_CONNECT_BASE_URL=https://fca.integ01.dev-agentconnect.fr/api/v2
      - CMD_AGENT_CONNECT_USER_INFO_SIGNED_RESPONSE_ALG=RS256
      - CMD_AGENT_CONNECT_SCOPES=openid given_name usual_name email
      - CMD_AGENT_CONNECT_ACR_VALUES=eidas1
```

run `docker compose build` then `docker compose up`.

TODO
- [ ] Fix error while calling session/end endpoint (cc @BenoitSerrano)
- [ ] Improve error handling
- [x] Plan production data migration
- [ ] Add the Agent Connect button in the UI (? cc @manuhabitela)

Commit:
✨(auth) support Agent Connect OIDC provider

This commit introduces a new authentication option within Hedge doc using
Agent Connect, which follows a stricter interpretation of the OIDC specification.

Added environment variables for Agent Connect (AC) client ID and client secret.
AC router is registered conditionally based on the presence of these environment
variables. BUT more environment variables are necessary, please have a look to
the docker compose file.

As AC supports a strict interpretation of the OIDC specification, we had to
implement a custom Passport.js Strategy. New routes were added to handle
the authorize, the authorize callback and the logout callback.

AC supports only the Authorization Code flow. It signs the `/userinfo` response.
It uses email for user lookup, and not the sub as recommended by the spec.
We managed user sessions with Passport.

While authorizing a user, we store its id token in its session data. While login out,
AC requires a parameter named `id_token_hint`, which is the `id_token`.

I relied on Panva library to validate the token and userinfo response. There is
room for improvement on these security checks. An iteration is necessary.

We should absolutely ensure the integrity and the authenticity of the requests and
token payloads.

Enhancement could be done on the error handling and logging strategy.

These changes ensure a robust integration of Agent Connect, facilitating secure,
efficient authentication and logout processes.

In production, users are identified by their email adresses. No data migration should
be required while switching from tchap id to AC.

Commit:
➕(deps) add openid-client dependency

The dependency was chosen by @BenoitSerrano. It will be necessary
to integrate Agent Connect OIDC provider.

Commit:
🔧(docker) fix docker compose build

User prompt was disabled for apt-get install, but not apgt-get update.
The command was prompting user permission to execute. Pass the same,
'-y' option, that anwser yes every time.

BenoitSerrano commented on app.js:
```diff
@@ -219,7 +219,9 @@ app.locals.authProviders = {
   email: config.isEmailEnable && !config.isEmailEnableForBotOnly,
-  allowEmailRegister: config.allowEmailRegister
+  allowEmailRegister: config.allowEmailRegister,
+  agentConnect: config.isAgentConnectEnable,
```
isAgentConnectEnabled*, non ?
Et la var d'env `agentConnect` est raccord avec ce qui est fait ailleurs dans l'app, mais c'est pas hyper explicite je trouve, par opposition à garder `isAgentConnectEnabled: config.isAgentConnectEnabled`

BenoitSerrano commented on lib/web/auth/agentConnect/index.js:
```diff
@@ -0,0 +1,194 @@
+    } = req.session
+
+    delete req.session.nonce
+    delete req.session.state
```
Le state doit être stockée dans la session pour la déconnexion non ?

BenoitSerrano commented on lib/web/auth/agentConnect/index.js:
```diff
@@ -0,0 +1,194 @@
+  const endSessionUrl = client.endSessionUrl({
+    id_token_hint: idToken,
+    post_logout_redirect_uri: config.serverURL + '/auth/agent-connect/logout/callback',
+    state
```
Voilà, je crois que le state doit être celui généré au moment de la connexion

Commit:
🩹(frontend) update mismatching html tag

Recent changes introduced by @manuhabitela were made on the UI.
One html closing tag wasn't matching. My bad I did a quick review.
Fixed it.
